csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -13,117 +13,334 @@
-    auto i_output = std::find_if(
-        outputs.begin(),
-        outputs.end(),
-        std::mem_fn(&TensorView::hasDeviceMesh));
-    if (i_output == outputs.end()) {
+    // All outputs of an expression (Welford, SDPA) should be uniformly sharded.
+    // We pick the most parallel output as the reference.
+    // This is to avoid picking seed/offset tvs in SDPA.
+    std::vector<TensorView*> sorted_outputs = sortTvsByDeviceDims(outputs);
+    
+    if (sorted_outputs.empty()) {
+      // No output with a device mesh.
       continue;
     }
-    TensorView* output_with_mesh = *i_output;
 
+    TensorView* ref_output = sorted_outputs.front();
+    NVF_ERROR(
+        ref_output != nullptr && ref_output->hasDeviceMesh(),
+        "Reference output ",
+        ref_output,
+        " has no device mesh.");
+
+    // For fusion inputs, only check if they have a device mesh. We do not
+    // modify their sharding. For non-fusion inputs, we try to propagate shardings
+    // from the reference output for parallel types that are not already present.
     const auto& inputs = ir_utils::filterByType<TensorView>(expr->inputs());
+    std::vector<TensorView*> sharding_candidates;
     for (auto* tv : inputs) {
-      if (!tv->hasDeviceMesh()) {
-        tv->setDeviceMesh(output_with_mesh->getDeviceMesh());
+      if (tv->isFusionInput()) {
+        if (!tv->hasDeviceMesh()) {
+          tv->setDeviceMesh(ref_output->getDeviceMesh());
+        }
+        continue;
+      }
+      if (!tv->hasDeviceMesh() || numDeviceDims(tv) == 0) {
+        sharding_candidates.push_back(tv);
       }
     }
+
+    if (sharding_candidates.empty()) {
+      continue;
+    }
+
+    int64_t did_pos = selectiveReorderDIDToFront(ref_output, {});
+    // Note: We do not have to manually shard for reshape here.
+    // TransformPropagator can handle reshapes when going from consumer to
+    // producer.
+    propagateDIDTransform(
+      /*ref=*/ref_output, 
+      /*tvs=*/sharding_candidates, 
+      /*did_pos=*/did_pos, 
+      /*allow_c2p=*/true, 
+      /*allow_p2c=*/false);
+    shardAllLike(ref_output, sharding_candidates);
   }
 
-  validateMeshes(fusion);
+  bool has_mesh = validateMeshes(fusion);
+  if (has_mesh) {
+    // Reorder the loop domain since the transform propagator may
+    // have reordered the iterdomains in loop domain. For example: Consider
+    // linear op: in = [b, m, k] weight = [DIDx(d), n/d, k] After
+    // transformation, the loop domain of linear output is [DIDx(d), n/d, b,
+    // m, r{k}]. Since, we set allocation to be the same as loop, we reorder it
+    // as allocation domain in the interim. Ideally, this should follow logical
+    // domain and DIDx axis at the front. The allocation domain should follow
+    // any stride order specified/inferred.
+    reorderLoopAsAllocation(fusion->allTvs());
+    for (auto tv : fusion->allTvs()) {
+      tv->setAllocationDomain(tv->getLoopDomain(), true);
```

Comment:
TODO: This seems to be only affecting `MHAFwd`. All other multidevice tests pass locally. Need to check the complete CI.

TODO: How to set contiguity here? (Follow the original contiguity and assign to derived ids)